### ISSUE TITLE
Help with BIOMOD_Modelling - [Ordinal abundance evaluation]

### ISSUE THREAD

**Issue by trehill** [biomodhub/biomod2#630]
**Context and question**
Hello, and thank you for your continued support.

I’m building ordinal abundance models for different species using biomod2. When I run models with only a training/calibration dataset, everything works fine. However, when I supply an independent evaluation (testing) dataset, the output from BIOMOD_Modeling is null.

I’ve previously built count-based abundance models for these same species using biomod2 without any issues, so this seems specific to the ordinal models. I’m also encountering the problem across multiple species, which makes me think it’s not a data-specific issue.

Could you clarify if this is expected behavior, or if I might be missing a step in properly formatting or specifying the evaluation data?

Many thanks in advance.

**Code used**
_**using only training dataset**_

>sp <- c('GUR')

 #define species to model (ch)
 >all_models <- c("CTA", "GLM","GAM", "MARS", "RF", "XGBOOST" ) #no GBM for ordinal data, no MAXENT for any abundance

>myBiomodData <- BIOMOD_FormatingData(resp.var = train_resp_df, #as SpatVector without associated data, for a single species
                                     expl.var = train_env, #SpatRaster containing explanatory variables in layers
                                     resp.name = sp, #character corresponding to species name
                                     resp.xy = train_xy, #if resp.var is vector, 2 column matrix associated xy coordinates
                                     dir.name = c('./models_biomod/ordinal/GUR'), 
                                     data.type = c("ordinal"), 
                                     na.rm = TRUE, 
                                     filter.raster = TRUE, #filter points in same cell
                                     seed.val = 1234
) 
>cv <- bm_CrossValidation(bm.format = myBiomodData,
                         strategy = "kfold", #options are random, k-fold, strat, block, env, or user-defined
                         nb.rep = 1, #integer, number of repetitions of CV points to be drawn 
                         k = 10, #number of partitions                                         
                         do.full.models = TRUE #calibrated and validated over the whole dataset (and pseudo-absence datasets) or not
) 

Checking Cross-Validation arguments...

   > k-fold cross-validation selection

>opt.d <- bm_ModelingOptions(data.type = 'ordinal', #character corresponding to data type to be used (see below)
                           models = all_models,
                           strategy = 'default',
                            bm.format = myBiomodData)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Build Modeling Options -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

> myBiomodModelOut <- BIOMOD_Modeling(bm.format = myBiomodData,
                                    models = all_models,
                                    CV.strategy = "user.defined",
                                    CV.user.table = cv,
                                    OPT.strategy = "user-defined",
                                    OPT.user = opt.d,
                                    metric.eval = c("Accuracy", "Precision", "F1"),
                                    var.import = 3,
                                    seed.val = 42,
                                    nb.cpu = 14)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Done -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

eval <- get_evaluations(myBiomodModelOut)

summary(eval)
full.name              PA                run                algo           metric.eval            cutoff     sensitivity   specificity   calibration    
 Length:198         Length:198         Length:198         Length:198         Length:198         Min.   : NA   Min.   : NA   Min.   : NA   Min.   :0.2740  
 Class :character   Class :character   Class :character   Class :character   Class :character   1st Qu.: NA   1st Qu.: NA   1st Qu.: NA   1st Qu.:0.3932  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character   Mode  :character   Median : NA   Median : NA   Median : NA   Median :0.5215  
                                                                                                Mean   :NaN   Mean   :NaN   Mean   :NaN   Mean   :0.5890  
                                                                                                3rd Qu.: NA   3rd Qu.: NA   3rd Qu.: NA   3rd Qu.:0.7240  
                                                                                                Max.   : NA   Max.   : NA   Max.   : NA   Max.   :1.0000  
                                                                                                NA's   :198   NA's   :198   NA's   :198                   
   validation       evaluation 
 Min.   :0.2270   Min.   : NA  
 1st Qu.:0.3668   1st Qu.: NA  
 Median :0.4940   Median : NA  
 Mean   :0.5140   Mean   :NaN  
 3rd Qu.:0.6560   3rd Qu.: NA  
 Max.   :0.8150   Max.   : NA  
 NA's   :18       NA's   :198 

_**using training AND evaluation dataset**_

>sp <- c('GUR')

 #define species to model (ch)
 >all_models <- c("CTA", "GLM","GAM", "MARS", "RF", "XGBOOST" ) #no GBM for ordinal data, no MAXENT for any abundance

>myBiomodData <- BIOMOD_FormatingData(resp.var = train_resp_df, #as SpatVector without associated data, for a single species
                                      expl.var = train_env, #SpatRaster containing explanatory variables in layers
                                      resp.name = sp, #character corresponding to species name
                                      resp.xy = train_xy, #if resp.var is vector, 2 column matrix associated xy coordinates
                                      dir.name = c('./models_biomod/ordinal/GUR'), 
                                      data.type = c("ordinal"), 
                                      eval.expl.var = eval_env, #SpatRaster without associated data, independent data to evaluate
                                      eval.resp.var = eval_resp_df, #SpatVector without associated data, independent data to evaluate 
                                      eval.resp.xy = eval_xy, #if resp.vect is vector, 2 column matrix of xy coordinates
                                      na.rm = TRUE, 
                                      filter.raster = TRUE, #filter points in same cell
                                      seed.val = 1234
 ) 

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= GUR Data Formating -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

      ! No data has been set aside for modeling evaluation
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Done -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
>cv <- bm_CrossValidation(bm.format = myBiomodData,
                         strategy = "kfold", #options are random, k-fold, strat, block, env, or user-defined
                         nb.rep = 1, #integer, number of repetitions of CV points to be drawn 
                         k = 10, #number of partitions                                         
                         do.full.models = TRUE #calibrated and validated over the whole dataset (and pseudo-absence datasets) or not
) 

Checking Cross-Validation arguments...

   > k-fold cross-validation selection

> opt.d <- bm_ModelingOptions(data.type = 'ordinal', #character corresponding to data type to be used (see below)
                            models = all_models,
                            strategy = 'default',
                            bm.format = myBiomodData)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Build Modeling Options -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

	>  CTA options (datatype: ordinal , package: rpart , function: rpart )...
	>  GLM options (datatype: ordinal , package: stats , function: glm )...
	>  GAM options (datatype: ordinal , package: mgcv , function: gam )...
	>  MARS options (datatype: ordinal , package: earth , function: earth )...
	>  RF options (datatype: ordinal , package: randomForest , function: randomForest )...
	>  XGBOOST options (datatype: ordinal , package: xgboost , function: xgboost )...

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= Done -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

>myBiomodModelOut <- BIOMOD_Modeling(bm.format = myBiomodData,
                                    models = all_models,
                                    CV.strategy = "user.defined",
                                    CV.user.table = cv,
                                    OPT.strategy = "user-defined",
                                    OPT.user = opt.d,
                                    metric.eval = c("Accuracy", "Precision", "F1"),
                                    var.import = 3,
                                    seed.val = 42,
                                    nb.cpu = 14)
! All models failed

>eval <- get_evaluations(myBiomodModelOut)

> summary(eval)
Length  Class   Mode 
     0   NULL   NULL 

```

``` 

**Environment Information**
Please paste the output of `sessionInfo()` in your current R session below.
```
# R version 4.5.0 (2025-04-11 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 26100)

Matrix products: default
  LAPACK version 3.12.1

locale:
[1] LC_COLLATE=English_Canada.utf8  LC_CTYPE=English_Canada.utf8    LC_MONETARY=English_Canada.utf8 LC_NUMERIC=C                   
[5] LC_TIME=English_Canada.utf8    

time zone: America/Vancouver
tzcode source: internal

attached base packages:
[1] parallel  splines   stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] corrplot_0.95        enmSdmX_1.2.12       kernlab_0.9-33       flexsdm_1.3.4        ggExtra_0.10.1       viridis_0.6.5        viridisLite_0.4.2   
 [8] CAST_1.0.3           doParallel_1.0.17    iterators_1.0.14     mop_0.1.2            dismo_1.3-16         raster_3.6-32        sp_2.2-0            
[15] tidyterra_0.7.1      caret_7.0-1          lattice_0.22-6       blockCV_3.1-5        ggsci_3.2.0          gridExtra_2.3        usdm_2.1-7          
[22] ggtext_0.1.2         xgboost_1.7.9.1      randomForest_4.7-1.2 maxnet_0.1.4         earth_5.3.4          plotmo_3.6.4         plotrix_3.8-4       
[29] Formula_1.2-5        gbm_2.2.2            mgcv_1.9-1           nlme_3.1-168         gam_1.22-5           foreach_1.5.2        mda_0.5-5           
[36] class_7.3-23         rpart_4.1.24         nnet_7.3-20          biomod2_4.3-2-2      cowplot_1.1.3        ggspatial_1.1.9      sf_1.0-20           
[43] terra_1.8-42         lubridate_1.9.4      forcats_1.0.0        stringr_1.5.1        purrr_1.0.4          readr_2.1.5          tibble_3.2.1        
[50] ggplot2_3.5.2        tidyverse_2.0.0      tidyr_1.3.1          dplyr_1.1.4         

loaded via a namespace (and not attached):
  [1] later_1.4.2            fields_16.3.1          R.oo_1.27.0            hardhat_1.4.1          pROC_1.18.5            xts_0.14.1            
  [7] lifecycle_1.0.4        globals_0.16.3         processx_3.8.6         MASS_7.3-65            magrittr_2.0.3         rmarkdown_2.29        
 [13] yaml_2.3.10            remotes_2.5.0          httpuv_1.6.15          spam_2.11-1            pkgbuild_1.4.7         DBI_1.2.3             
 [19] maps_3.4.2.1           abind_1.4-8            R.utils_2.13.0         ipred_0.9-15           lava_1.8.1             listenv_0.9.1         
 [25] units_0.8-7            parallelly_1.43.0      PresenceAbsence_1.1.11 codetools_0.2-20       gstat_2.1-3            xml2_1.3.8            
 [31] tidyselect_1.2.1       farver_2.1.2           stats4_4.5.0           jsonlite_2.0.0         e1071_1.7-16           survival_3.8-3        
 [37] systemfonts_1.2.2      tools_4.5.0            twosamples_2.0.1       ragg_1.3.3             snow_0.4-4             Rcpp_1.0.14           
 [43] glue_1.8.0             prodlim_2024.06.25     xfun_0.52              withr_3.0.2            fastmap_1.2.0          callr_3.7.6           
 [49] digest_0.6.37          timechange_0.3.0       R6_2.6.1               mime_0.13              textshaping_1.0.0      colorspace_2.1-1      
 [55] R.methodsS3_1.8.2      intervals_0.15.5       generics_0.1.3         data.table_1.17.0      recipes_1.2.1          FNN_1.1.4.1           
 [61] ModelMetrics_1.2.2.2   pkgconfig_2.0.3        gtable_0.3.6           timeDate_4041.110      htmltools_0.5.8.1      dotCall64_1.2         
 [67] automap_1.1-16         scales_1.3.0           spThin_0.2.0           doSNOW_1.0.20          gower_1.0.2            knitr_1.50            
 [73] rstudioapi_0.17.1      tzdb_0.5.0             reshape2_1.4.4         spacetime_1.3-3        curl_6.2.2             zoo_1.8-13            
 [79] proxy_0.4-27           KernSmooth_2.23-26     miniUI_0.1.1.1         desc_1.4.3             pillar_1.10.2          grid_4.5.0            
 [85] reshape_0.8.9          vctrs_0.6.5            promises_1.3.2         xtable_1.8-4           evaluate_1.0.3         cli_3.6.4             
 [91] compiler_4.5.0         rlang_1.1.5            crayon_1.5.3           future.apply_1.11.3    labeling_0.4.3         classInt_0.4-11       
 [97] ps_1.9.0               plyr_1.8.9             stringi_1.8.7          stars_0.6-8            munsell_0.5.1          Matrix_1.7-3          
[103] hms_1.1.3              patchwork_1.3.0        future_1.40.0          shiny_1.10.0           gridtext_0.1.5  
```

**Comment by MayaGueguen** [comment]
Hello there,

Thank you for posting 🙏 

I'm not sure actually it has to do with the ordinal data, rather than with the evaluation dataset 👀 
Indeed, you see that when calling the [BIOMOD_FormatingData](https://biomodhub.github.io/biomod2/reference/BIOMOD_FormatingData.html) function, you get the message : 

>   ! No data has been set aside for modeling evaluation

meaning that it does not seem to understand / take into account the evaluation data.
Could you please send me the output of the `BIOMOD_Report(bm.out = myBiomodData, strategy = "report")` command ?

Maya

**Comment by trehill** [comment]
Hi Maya. 

Oh- I see. I was working on using the abundance data for continuous abundance data and was receiving the same warning but subsequent functions seemed to be working okay. Here is the output of BIOMOD_Report(). Thank you for all your help!


[biomod2_report_GUR.zip](https://github.com/user-attachments/files/20041200/biomod2_report_GUR.zip)

**Comment by HeleneBlt** [comment]
Hi ! 

You were right to ignore this message, it was a small error 😬  It's fixed now! 

We have also sent a commit to correct the formatting of the evaluation data with ordinal data. So if you update to the new version, everything should work fine. 
However, we are still working on the predictions with the ensemble model with ordinal data. 
So you can expect some changes (for the better) soon. 🤞 


Hélène
